feat(new transform): (key/value) kv parser

Nick011 opened the pull request:
# Abstract

I work at a company that currently ingests a large amount of data. I'd like to reduce our spending on hardware. Logstash is fine for smaller amounts of traffic and most projects. However, when ingesting data and attempting to parse at scale, Logstash's performance is not sufficient.

Unfortunately, we've built up quite a bit of technical debt around Logstash. To transition to a more performant solution, such as vector, we will have to do it piecemeal. To easily accomplish this, we will need more feature parity with Logstash than Vector currently provides.

In that vein, I'd like to present *kv_parser*, the first of possibly multiple PRs to bring additional feature parity to Logstash. This transforms attempts to provide similar functionality to [Logstash's KV Paser](https://www.elastic.co/guide/en/logstash/current/plugins-filters-kv.html)

# Help

So, I've made this work to the point where I've actually tested this against streaming production data over HTTP. I need help with the following:

- I am a beginner Rust developer. 10+ years dev exp with python, js, and other languages. This Rust attempt comes after reading "the book", doing the rustling course, and one other 300 line side project.
- For some reason, it doesn't seem to accept square brackets for keys.
- I couldn't figure out docusaurus. TBH, I spent about 20 mins max on it. We use Hugo. I ain't got time for yet another static site generator. Would be happy to document if it was just markdown and obvious where to put it.

# Example

Here's an example `toml` config with test displaying why this transform is needed:

```
[sources.prod_logstash]
  type = "http"
  address = "0.0.0.0:80"
  encoding = "json"
  #max_length = 1_024_000_000
  #mode = "tcp"


[transforms.checkpoint_parser]
  drop_field = true
  field = "checkpoint_data"
  field_split = ":"
  inputs = ["prod_logstash"]
  separator = ";"
  target_field = "data"
  trim_key = "\""
  trim_value = "\""
  type = "kv_parser"

  types.flags = "int"
  types.logid = "int"
  types.sequencenum = "int"
  types.time = "int"
  types.version = "int"
  types.__nsons = "int"
  types.__p_dport = "int"
  types.__pos = "int"
  types.bytes = "int"
  types.client_inbound_bytes = "int"
  types.client_inbound_packets = "int"
  types.client_outbound_bytes = "int"
  types.client_outbound_packets = "int"
  types.server_outbound_bytes = "int"
  types.server_outbound_packets = "int"
  types.start_time = "int"


[sinks.elastic]
  type = "elasticsearch" # required
  inputs = ["pipeline_steps"] # required
  compression = "none" # optional, default
  healthcheck = true
  host = "http://x.x.x.x:9200" # optional, no default
  index = "%Y-%m-%d-{{ event_type }}-records"

  auth.strategy = "basic"
  auth.user = "xxx_user"
  auth.password = "xxxxxxxx"

  buffer.type = "memory"
  buffer.timeout_secs = 30


[[tests]]
  name = "test checkpoint parser"

  [[tests.inputs]]
    insert_at = "checkpoint_parser"
    type = "log"
    log_fields.checkpoint_data = """
     action:\"Accept\"; flags:\"802832\"; ifdir:\"inbound\"; ifname:\"eth2-05\"; logid:\"6\"; loguid:\"{0x5f0fa4d6,0x1,0x696ac072,0xc28d839a}\"; origin:\"x.x.x.x\"; originsicname:\"CN=MinusMorgul,O=x.xx.lx.x\"; sequencenum:\"51\"; time:\"1594860763\"; version:\"5\"; __nsons:\"0\"; __p_dport:\"0\"; __policy_id_tag:\"product=VPN-1 & FireWall-1[db_tag={70434369-839D-0047-8E41-26B16C9B99D6};mgmt=CPManage02.poet.local;date=1594324415;policy_name=Edge_Firewalls\\]\"; __pos:\"7\"; bytes:\"557\"; client_inbound_bytes:\"345\"; client_inbound_interface:\"eth2-05\"; client_inbound_packets:\"5\"; client_outbound_bytes:\"212\"; client_outbound_packets:\"10\"; context_num:\"4294967295\"; elapsed:\"0\"; hll_key:\"13860728580460784776\"; packets:\"15\"; product:\"Log Update\"; segment_time:\"1594860758\"; server_inbound_bytes:\"212\"; server_inbound_packets:\"5\"; server_outbound_bytes:\"345\"; server_outbound_interface:\"eth1-02\"; server_outbound_packets:\"10\"; start_time:\"1594860758\";"
    """

    [[tests.outputs]]
      extract_from = "checkpoint_parser"
      [[tests.outputs.conditions]]
        type = "check_fields"
        "data.action.eq" = "Accept"


```

Closes: https://github.com/timberio/vector/issues/3283

Nick011 committed:
feat(transforms): Add kv parser

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
feat(transforms): Add kv parser

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Fix kv tests

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Fix kv tests

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
feat(kv_parser): add target_field and trim args

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
feat(kv_parser): add target_field and trim args

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add target_field and trim args to tests

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add target_field and trim args to tests

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
use same var name for value

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
use same var name for value

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore: add tests for trim option

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore: add tests for trim option

Signed-off-by: Nick Hagianis <nhagianis@gmail.com>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore: run fmt

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore: run fmt

Signed-off-by: Nick <nhagianis@gmail.com>

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: String,
```
Given it's usage, this would probably be better as an `Option<String>` where `None` represents splitting on white space.

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+
+    if field_split.is_empty() {
+        let mut kv_pair = pair.split_whitespace();
+        let key = kv_pair.nth(0)?;
+        let val = kv_pair.nth(1)?;
+        let count = kv_pair.count();
+
+        if count < 2 {
+            return None;
+        } else if count > 2 {
+            error!(
+                message = "KV parser expected 2 values, but got {count}",
+                count = count,
+                rate_limit_secs = 30
+            );
+            return None;
+        }
```
```suggestion
        let key = kv_pair.next()?;
        let val = kv_pair.next()?;
        if let kv_pair.next().is_some() {
            error!(
                message = "KV parser saw more than one separator",
                rate_limit_secs = 30
            );
            return None;
        }
```

Nick011 replied:
This is what I came here for, good suggestions by more experienced rust devs!

Couple questions:
- is `next()` more performant or idiomatic than `nth()`, or just better in this case because it eliminates the need for `count`?
- the error case of more than one separator _should_ be impossible here but, I figured it would be safest to write a catch for it. The other fail case of having no separator would be possible and more common. Should we add an `error` for that so it's logged?

bruceg replied:
I wouldn't expect a major performance difference in this case. However, `next()` is the underlying iterator method, and `nth()` is a wrapper, so I don't think the latter can produce _better_ code in this case (might not be true for iterators over actual arrays or ranges). In terms of eliminating the count, that eliminates the need to continue the iteration beyond the next match.

I don't see the case of more than one separator being eliminated through code constructs, so it has to be handled. Even if you think it should never appear in the input, it is still a possibility.

It is probably worth answering the question if `key=value=another` should be an error or parsed as `key` and `value=another`.

Nick011 replied:
I think I would prefer if this to be parsed `key: "value=another"`. There are arguments I can think of on both sides. I don't have really strong feelings either way.

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+    trim_value: &Option<String>,
+) -> Option<(Atom, String)> {
+    let pair = pair.trim();
+    let mut field;
+
+    if field_split.is_empty() {
```
This would be better written as `let field = if field.split.is_empty()` with the last line of each block being the new value for `field`.

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+
+        field = [key, val];
+    } else {
+        let split_index = pair.find(field_split).unwrap_or(0);
```
It might be better to `return None` here instead, possibly with an error.

lukesteensen reviewed (commented):
This looks really interesting, thanks @Nick011!

It seems like a more flexible version of the [`logfmt` parser](https://vector.dev/docs/reference/transforms/logfmt_parser/), so I wonder if it'd make sense to lump them together somehow. I'm curious what @binarylogic thinks.

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+        let split_index = pair.find(field_split).unwrap_or(0);
+        let (key, val) = pair.split_at(split_index);
+        let key = key.trim();
+        let val = val.trim_start_matches(field_split).trim();
```
```suggestion
        let val = pair[split_index + len(field_split)..].trim();
```
It's too bad Rust's strings don't have a built-in split that handles this internally.

Nick011 replied:
Yeah, I couldn't find anything. Kind of sad to hear this is the best we can do with the standard lib.

I'm guessing this is more performant than `trim`?

bruceg replied:
I'm not sure what you're asking about `trim`, since it's still included. However note that the `trim*` functions remove _all matches_ rather than just the first, so if you want to only remove one they are not appropriate.

Strings are indexed at byte offsets not character indices, so your second suggestion is correct. I still sometimes think `len(X)` from long use of Python.

The use of `split_at` with a separator isn't particularly useful, as it will always require post-processing to remove the separator.

Nick011 replied:
Sorry, I was talking about `trim_start_matches` and the diff here. As far as I could tell, they both do the same thing? If that's not the case, can you point me to some docs or elaborate? If it is, what are the other reasons for the suggestion?

bruceg replied:
`trim_start_matches` may strip multiple matches. So, if the separator is `=` and the string is `key===value` then `val` will contain `value` with `trim_start_matches`, but `==value` with the manual split. This is an important difference.

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+    }
+
+    for (i, trim) in [trim_key, trim_value].iter().enumerate() {
+        if let Some(trim) = trim {
+            let trim_chars: Vec<char> = trim.chars().collect();
```
Since this will be constant, it should be computed when `struct KeyValue` is created rather than once for each line.

Nick011 replied:
Yup, great catch, that should increase perf and reduce memory usage a bit. I'll address this when I refactor to move the `kv_parser` function to a method.

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+    for (i, trim) in [trim_key, trim_value].iter().enumerate() {
+        if let Some(trim) = trim {
+            let trim_chars: Vec<char> = trim.chars().collect();
+            field[i] = field[i].trim_matches(&trim_chars as &[_]);
```
…and then this can simply be `self.trim_key.as_ref().map(|trim| field[0] = field[0].trim_matches(trim));` (repeat for `trim_value`)

bruceg commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+}
+
+fn kv_parser(
+    pair: String,
```
Since `pair` isn't modified, it can be type `&str` instead so you don't need to create a string when calling it.

binarylogic commented:
Thanks @Nick011, agree with @lukesteensen. If we could expand this to cover the basic parsing that the ‘logfmt’ transform supports then we can deprecate that in favor of this. What do you think? We could also follow up with that in a separate PR.

Nick011 committed:
ref(kv_parser): rename file to be more explicit

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
ref(kv_parser): rename file to be more explicit

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
ref(kv_parser): convert field_split to opt string

- Also changed default to equal sign

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+use super::Transform;
```
Yeah, it's probably better to be explicit.

Nick011 commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+    #[derivative(Default(value = "true"))]
+    pub drop_field: bool,
+    pub field: Option<Atom>,
+    #[derivative(Default(value = " "))]
+    pub field_split: String,
```
I wasn't really sure what the default should be or if there should even be a default here.

> Is space-separated more common usage?

I think this is the right idea though, and I think it should be `=`. The other common option being `:`. If we want to consider combining this with `logfmt` as mentioned by @lukesteensen and @binarylogic  then `=` is the obvious choice.

Nick011 committed:
fix module rename

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add behavior test

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented on src/transforms/kv_parser.rs:
```diff
@@ -0,0 +1,404 @@
+    field_split: &String,
+    trim_key: &Option<String>,
+    trim_value: &Option<String>,
+) -> Option<(Atom, String)> {
```
I've made the change but, `make check-clippy` didn't complain about anything. Nor did my IDE. 

```
vector on  feat/kv-parser [$!] is 📦 v0.11.0 via 🦀 v1.45.0 on ☁️  us-east-1
❯ make check-clippy
cargo clippy --workspace --all-targets -- -D warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
```

Maybe I have something configured incorrectly?

>this should probably be moved into a method of that struct

I agree. I'll work on that.

bruceg replied:
Hmm, there is [a lint for it](https://rust-lang.github.io/rust-clippy/master/#ptr_arg), but we currently allow it.

Nick011 committed:
use let..if syntax instead of mut field

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
use let..if syntax instead of mut field

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
apply a couple suggestions from PR review

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
apply a couple suggestions from PR review

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
Thanks for the review @bruceg! I learned some things about Rust, which is one of my goals here. Regardless of the decision on `logfmt` merging, I will complete the rest of the suggestions. They're all great.

@binarylogic and @lukesteensen I think it makes sense to consider combining `logfmt` with a more configurable parser like `key_value` but, I have some UX concerns.

- `logfmt` parser currently supports a leading string before splitting. This would create an added layer of difficulty to attempt to support in my KV implementation. `"message": "Hello world duration=2s user-agent=\"Firefox/47.3 Mozilla/5.0\""` would result in the first key being `Hello world duration` in my implementation. While, `logfmt` is smart and will strip that down to `duration`. In logstash I'm currently handling that case with another (previous) transformation. I deal with a lot of syslog and logs from hardware appliances such as switches and firewalls. Often times, what we will do is first split on a certain character, or do a `grok` to get the KV data and then send just that data into the KV parser. So, something like `Hello world - duration=2s user-aget=firefox ...` we would just split on `-` or do a grok `{lead_data} - {xyz-firewall-data}` and then run `xyz-firewall-data` through the kv.
- Ideally we would maintain a backwards compatibility as to not disrupt users who have implemented `logfmt`
   - While, it should be pretty simple to just say; if the separator is whitespace and the splitter is `=` run it through `logfmt`, this creates a documentation situation where you have to explain that; if you have this particular combination of options, it will trim the leading string up to the first key but, if not this exact combination of options, you have to handle that yourself.
   - I don't know if that is painful enough to be a blocker. For me; I'd prefer to not attempt to implement that logic for other KV formats.
- Barring the above being a show stopper and figuring out a solution there; how would you like to see implementation? Would you prefer I remove the `logfmt` parser, roll the logic into the `key_value_parser` and create an alias (not sure how to do that, would rely on suggestions from @bruceg) to not introduce a breaking change? Or, would you prefer I call the current `logfmt` transformer from the `key_value_transform` when the case matched? Or, do you want a breaking change and have `logfmt` nuked from orbit?

Nick011 commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,405 @@
+            None,
+        );
+
+        // println!("{}", log.keys().fold("".to_string(), |res, key| [res, key].join(", ")));
+        // Review: This fails and I don't know why. The key is missing from the results.
+        // assert!(log.contains(&"[score]".into()));
```
@bruceg Any ideas why `[` chars would cause problems? As far as I can tell, the parser I've written should be okay with this char. This was a very weird problem. I think I spent an hour or two and couldn't figure out why it would strip keys with`[]`.

bruceg replied:
Our internal event field name parsing uses `[X]` to mean an index on an array element, and so is trying to handle `[score]` as that and failing.

Nick011 replied:
I've removed this test.

binarylogic commented:
@Nick011 thanks for outlining everything. Given your comments, I vote that we handle deprecating the `logfmt` transform in a separate PR. Let's get this to a shippable state that resolves your concerns, then we can work on merging them.

binarylogic commented:
@Nick011 when you're ready for another review would you mind taking the PR out of draft?

Nick011 committed:
ref: move kv_parse func to method

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
ref: move kv_parse func to method

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
delcare key_value_parser in config

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Add key_value to internal_events

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
run fmt

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
run fmt

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add default separator

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix lint errors

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix key_value parser name in test

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
I've updated the PR based on feedback. It will be easier to handle merging `logfmt` in another PR. I definitely think it should happen.

I created a file for `key_value` in `/internal_events/` based off the `json` file. I see where yall are going there but, I'm not sure if I did everything (right).

bruceg commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,409 @@
+            (key, val)
+        };
+
+        let key = if let Some(trim_key) = &self.trim_key {
+            fields.0.trim_matches(trim_key as &[_])
+        } else {
+            fields.0
+        };
```
While this can be done with `map` and `unwrap_or`, I think a `match` is best here:
```suggestion
        let key = match &self.trim_key {
            Some(trim_key) => fields.0.trim_matches(trim_key as &[_]),
            None => fields.0,
        };
```

bruceg commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,409 @@
+            fields.0
+        };
+
+        let val = if let Some(trim_value) = &self.trim_value {
+            fields.1.trim_matches(trim_value as &[_])
+        } else {
+            fields.1
+        };
```
Ditto match

Nick011 replied:
It seems I found an `if let Some` hammer and over used it on my last round of updates, haha. Ditto for the previous Ditto reply; what are some things you look at and take into consideration what making decisions in these situations where there is more than one way to do it in Rust?

bruceg replied:
I tend to use a few relatively simple metrics to guide me:
1. What is the highest-level language construct for this task? (functional operators >= `match` >= `let X = if … else …` >= `let mut X; if _ { X = … } else { X = … }`)
1. What construct most clearly represents what I want to achieve (as opposed to what the compiler will translate it into)?
1. What results in the most compact representation of the semantics?

For `if let` in particular, if it is balanced by an `else` clause, I will almost always use `match` instead as it better represents what is being done (there are a few cases where `rustfmt` will turn the `if` into a single line but the `match` will always be multi-line). Similarly for pattern matching, using `match` across all the values that need matching lets us be sure all the cases are enumerated, and the compiler is free to optimize it into nested `if/else` blocks internally. Further, if the operation can be represented by functional operators (`.map`, `.and_then`, etc), I will tend to prefer those. In this case it would be possible to use them, but it doesn't clarify anything:
```rust
    let val = self
        .trim_value
        .as_ref()
        .map(|trim_value| fields.1.trim_matches(trim_value as &[_]))
        .unwrap_or(fields.1);
```

Nick011 replied:
This is awesome, and very helpful, thanks!

bruceg commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,409 @@
+            let key = key.trim();
+            let val = pair[split_index + field_split.len()..].trim();
+
+            if key.is_empty() || val.is_empty() {
+                return None;
+            }
```
Is `"key="` (empty value) valid here?

Nick011 replied:
Currently, that key would be dropped, but the transform would succeed with all other key/values. [There's a test for it](https://github.com/timberio/vector/pull/3284/files#diff-cb98402adc309a1927a810f3e4d88de7R334). 

So yes, the following message would be valid:
`name=Nick, age=, job=software`

Would result in:
```
{
  "name": "Nick",
  "job": "software"
}
```

This was intentional. I'm open to whatever you think is best but, I would prefer the whole transform not fail on this scenario where it's _mostly_ parse-able.

bruceg replied:
Right, the question is if this should result in `"age": ""` or `"age": null`. I personally think that would be less surprising because it doesn't throw away incoming data (the key), but I'm not the target user for this transform.

Nick011 replied:
Ahh, that is a good point. I agree, but I'm not sure what to go with. I'm leaning `null` because an empty string implies type. I'll move in that direction but, open to opinions or changing it in the future.

Nick011 replied:
Empty string was easier. I don't know how much it's going to matter. I'm using Elasticsearch for a sink so, I'll see what happens there if it creates problems.

bruceg replied:
I'd tend towards `null` as well, but if empty string is easier then go with that. We can figure out how best to change the behavior if it actually causes a problem with real data.

bruceg reviewed (changes requested):
Thanks for adding the internal events. You are right we are pushing to add them for everything, so that's great. I don't see them being emitted anywhere, though. This is getting pretty close otherwise.

Nick011 committed:
remove atom type

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/key_value.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/transforms/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix value conversion

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/mod.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/transforms/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
update trim parameters to collect using map

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,409 @@
+            Some(val.chars().collect())
+        } else {
+            None
+        };
```
Other than the obviously better aesthetics of `map`, are there differences in what is happening here?

bruceg replied:
No, it wouldn't change what is happening. It simply better matches the actual operation happening—you're mapping the value inside the option to a new value, hence map.

Nick011 committed:
Update src/transforms/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
convert if let to match for trimming

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
convert if let to match for trimming

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix internal events to provide some info

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
Tried to update the transform to trigger at least a few basic internal events. I'm not really sure if I'm doing this right or how to test it.

bruceg commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,397 @@
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
```
I think I missed seeing this before. This should only be true if `field != target_field`:
```rust
        let drop_field = self.drop_field && match &self.target_field {
            Some(target_field) => field != target_field,
            None => true,
        };
```
or
```rust
        let drop_field = self.drop_field && self.target_field.as_ref().map(|target_field| field != target_field).unwrap_or(true);
```
This is to avoid dropping the target that was just parsed.

Nick011 replied:
Yes, this is a great check. Thanks!

bruceg reviewed (changes requested):
Looks better!

Nick011 committed:
change null val to empty string instead of drop

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add defaults for separator and field_split

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix naming convention for KeyValueParseFailed

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix internal events

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
ensure drop_field can't drop the target_field

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
run fmt

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add meta for kv parser

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Add example to meta

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Run linter

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
Ready for more review @bruceg 

- Responded to all comments on previous feedback
- Added behavior tests
- Added documentation in the `.meta` dir
- Added to `CODEOWNERS`

Nick011 committed:
add CODEOWNERS for key value parser

Signed-off-by: Nick <nhagianis@gmail.com>

bruceg reviewed (changes requested):
Very close! Thanks for sticking with this.

Nick011 committed:
fix docs for running benchmarks

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
Update src/internal_events/key_value_parser.rs

Co-authored-by: Bruce Guenter <bruce@untroubled.org>
Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
add internal event for multiple white space split results

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
No prob, there are numerous aspects about this that are cool to me. I just don't have a lot of time. I also went on vacation for a week, so.

Nick011 commented on src/transforms/key_value_parser.rs:
```diff
@@ -0,0 +1,400 @@
+                error!(
+                    message = "KeyValue parser saw more than one separator",
+                    rate_limit_secs = 30
+                );
```
I almost removed this functionality. It's a weird-ish edge case to want to split a field on white space. I considered converting empty string inputs to the default `=`. I figured it was probably best to not restrict a user if they _really_ wanted  to do this for some reason.

bruceg replied:
If splitting a key+value on white space is that low value, it would be best to remove this as the default. I don't see it being useful myself. Defaulting to `"="` as you suggest seems to make more sense.

Nick011 replied:
I've updated it to remove whitespace as an option for field_split and default to `=`.

Nick011 committed:
run fmt

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
remove field splitting on whitespace

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
@bruceg what do you think?

bruceg commented:
@lukesteensen @JeanMertz I've looked at this so many times, the details are getting fuzzy. Could one of you take a second look before I merge it (and @binarylogic for the docs)?

lukesteensen reviewed (approved):
This looks good to me! Thanks again @Nick011 and @bruceg for shepherding this through.

The one thing that jumps out to me that'd prevent this from superseding the `logfmt` parser is handling quoted strings and escaping, but I think that's something we can do in a followup once we have a plan to merge and deprecate. We can open an issue to discuss further.

Otherwise, I'll defer to @binarylogic for the docs.

jamtur01 commented:
@binarylogic Ping re the docs for this.

@Nick011 Thanks for your patience! We just recently moved from the TOML docs to Cue-based docs (you can see an example here: https://github.com/timberio/vector/blob/master/docs/reference/components/transforms/grok_parser.cue) so we just need to update the docs to get this merged.

Nick011 committed:
chore: remove Atom and implement async build method

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
fix(kv tests): convert tests to async

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
WIP: new style docs

Signed-off-by: Nick <nhagianis@gmail.com>

jamtur01 commented:
@Nick011 A tiny number of remaining Cue errors:

```
Run make check-docs
./scripts/check-docs.sh
Validating docs/**/*.cue formatting.
expected label or ':', found '{':
    ./docs/reference/components/transforms/key_value_parser.cue:94:16
expected label or ':', found '{':
    ./docs/reference/components/transforms/key_value_parser.cue:106:12
expected label or ':', found '{':
    ./docs/reference/components/transforms/key_value_parser.cue:116:14
unknown escape sequence:
    ./docs/reference/components/transforms/key_value_parser.cue:118:100
unknown escape sequence:
    ./docs/reference/components/transforms/key_value_parser.cue:119:85
make: *** [Makefile:812: check-docs] Error 1
Error: Process completed with exit code 2.
```

You can also check locally with make check-docs (having Cue installed on the host).

Thanks! Almost there.

Nick011 committed:
chore(kv docs): add examples and description

Signed-off-by: Nick <nhagianis@gmail.com>

jamtur01 commented:
@Nick011 Looks like a couple of issues still. You don't need to force push BTW - we squash and merge.

Nick011 committed:
chore(kv docs): delete old style docs

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore(kv docs): run check docs

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore(kv parser): update to latest framework

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 committed:
chore(kv tests): fix behaviour tests

Signed-off-by: Nick <nhagianis@gmail.com>

Nick011 commented:
@jamtur01 there are several tests failing for me locally unrelated to the key value parser. I think I covered docs and updates required by the framework since the last time I touched this. Tests for key value parser are passing locally.

> You don't need to force push BTW - we squash and merge.
Is the work flow to merge changes from master locally rather than rebase? I'll do that from now on. Part of it is that I always forget to `signoff` as well.

jamtur01 commented:
Hi @Nick011 - some tinsy failures left:

```
5
components.transforms.key_value_parser.how_it_works.stuff: conflicting values "\tThe Key Value Parser accepts structured data that can be split on a character, or group of characters, and extracts it into a\n\tjson object (dictionary) of key/value pairs. The `separator` option allows you to define the character(s) to perform the initial\n\tsplitting of the message into pairs. The `field_split` option allows you to define the character(s) which split the key from the value." and {name:Name,title:string,body:string,sub_sections?:[...{title:string,body:string}]} (mismatched types string and struct):
216
    ./docs/reference.cue:462:35
217
    ./docs/reference/components/transforms/key_value_parser.cue:143:10
218
make: *** [Makefile:824: check-docs] Error 1
219
Error: Process completed with exit code 2.
```

Nick011 commented:
Yeah, I saw that right after I posted. My bad.

Nick011 committed:
chore(kv docs): fix how_it_works section

Signed-off-by: Nick <nhagianis@gmail.com>

jamtur01 commented:
```
components.transforms.key_value_parser.configuration.target_field.type.string.default: incomplete value string | null
components.transforms.key_value_parser.configuration.trim_key.type.string.default: incomplete value string | null
components.transforms.key_value_parser.configuration.trim_value.type.string.default: incomplete value string | null
components.transforms.key_value_parser.features.parse.format.url: incomplete cause disjunction
```

@Nick011 Couple of minor issues and we're there. :)

Nick011 committed:
fix(kv docs): set default to null

Signed-off-by: Nick <nhagianis@gmail.com>

jamtur01 commented:
@Nick011 Can you rebase this? The remaining errors are cleared up in master!

Nick011 commented:
@jamtur01 Did a rebase. Looks like there's still a couple unrelated things failing. Let me know if there is anything else you want me to address

jamtur01 committed:
Merge branch 'master' into feat/kv-parser

jamtur01 committed:
Merge branch 'master' into feat/kv-parser

jamtur01 commented:
@Nick011 I cleared the unrelated formatted error but I see a Cue fmt error. I am going to merge it anyways given how patient you've been and fix it myself once the other tests run.

jamtur01 commented:
Thanks for your contribution @Nick011 - this is now merged! 👍 🥇 🥂 🎆

Nick011 commented:
Thank you @jamtur01 , @bruceg , @binarylogic , and @lukesteensen ! 

I know you all invested quite a bit helping me along. I hope this feature brings value to your business and users, as well as opens up additional use cases for vector. I wish I had more experience with rust and required less help. I'd like to especially thank @bruceg for his multiple reviews, and explanations of rust idioms throughout this process, (I really learned a lot).

While [things have changed rather quickly over the past couple months](https://www.connectwise.com/announcements?utm_source=perch&utm_medium=webref&utm_campaign=itncannouncement), I still intend to advocate for the use of vector in our stack. You all have built an amazingly innovative technology. Thank you for letting me be a part of it. I will be around to support this module as much as I can. Feel free to contact me directly nhagianis@gmail.com.

I'm rooting for timber.io and wish you all great success!

zhoujw8792 commented:
@Nick011 I've been looking forward to this feature,when vector-0.11.0 running windows 10 and centos 7.3 tests, The following problems were found. when values contains separator character(s) ,value occur data loss(https://github.com/timberio/vector/issues/5347)
